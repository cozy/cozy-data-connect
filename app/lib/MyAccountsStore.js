/* global fetch */
/* global cozy */
import { Component } from 'preact'

const INSTALL_TIMEOUT = 120

export default class MyAccountsStore {
  constructor (connectors, folders, context) {
    this.listener = null
    this.connectors = connectors
    this.folders = folders
    this.useCases = require(`../contexts/${context}/index`).useCases
  }

  subscribeTo (connectorId, listener) {
    this.listener = listener
    return this.find(c => c.id === connectorId)
  }

  unsubscribe () {
    this.listener = null
  }

  updateConnector (connector) {
    this.connectors = this.connectors.map(
      c => c.slug === connector.slug ? Object.assign({}, c, connector) : c
    )
    if (this.listener) {
      this.listener(this.find(c => c.slug === connector.slug))
    }
  }

  getCategories () {
    return this.connectors.map(a => a.category).filter((cat, idx, all) => all.indexOf(cat) === idx)
  }

  getUseCases () {
    return this.useCases
  }

  find (cb) {
    return this.connectors.find(cb)
  }

  findConnected () {
    return this.connectors.filter(c => c.accounts.length !== 0)
  }

  findByCategory ({filter}) {
    return filter === 'all' ? this.connectors
      : this.connectors.filter(c => c.category === filter)
  }

  findByUseCase (slug) {
    let useCase = this.useCases.find(u => u.slug === slug)
    return useCase.connectors.map(c1 => this.find(c2 => c1.slug === c2.slug))
  }

  connectAccount (connectorId, values, accountId = 0) {
    let connector = this.find(c => c.id === connectorId)
    connector.accounts[accountId] = values
    return this.importConnector(connector)
      .then(() => this.refreshFolders())
      .then(() => this.startConnectorPoll(connector.id))
  }

  addAccount (connectorId, values = {}) {
    let connector = this.find(c => c.id === connectorId)
    connector.accounts.push(values)
    this.updateConnector(connector)
  }

  updateAccount (connectorId, accountIdx, values) {
    let connector = this.find(c => c.id === connectorId)
    connector.accounts[accountIdx] = values
    return this.putConnector(connector)
  }

  synchronize (connectorId) {
    let connector = this.find(c => c.id === connectorId)
    return this.importConnector(connector)
      .then(() => this.startConnectorPoll(connector.id))
  }

  deleteAccount (connectorId, accountIdx) {
    let connector = this.find(c => c.id === connectorId)
    connector.accounts.splice(accountIdx, 1)
    return this.importConnector(connector)
      .then(() => this.updateConnector(connector))
  }

  fetchOrInstallConnector (slug) {
    const connectorSpec = this.connectors.find(conn => conn.slug === slug)
    return this.fetchConnector(connectorSpec)
    .then(connector => {
      if (connector === null) {
        return this.installConnector(connectorSpec)
      } else {
        return connector
      }
    })
  }

  // Fetch stack data for the given connector
  fetchConnector (connector) {
    return cozy.client.data.defineIndex('io.cozy.konnectors', ['slug'])
      .then(index => cozy.client.data.query(index, {selector: {slug: connector.slug}}))
      .then(list => {
        if (list.length) {
          return list[0]
        } else {
          return null
        }
      })
      .then(connector => {
        if (connector) this.updateConnector(connector)
        return connector
      })
  }

  // Trigger the installation of the given connector
  // and check that the installation is terminated
  installConnector (connector) {
    return this.fetch('POST', `/konnectors/${connector.slug}?Source=${encodeURIComponent(connector.repo)}`)
      .then(() => new Promise((resolve, reject) => {
        const idTimeout = setTimeout(() => {
          reject(new Error('konnector installation timeout error'))
        }, INSTALL_TIMEOUT * 1000)

        // monitor the status of the connector
        const idInterval = setInterval(() => {
          this.fetchConnector(connector)
            .then(connector => {
              if (connector.state === 'ready') {
                clearTimeout(idTimeout)
                clearInterval(idInterval)
                resolve(connector)
              }
            })
            .catch(err => {
              clearTimeout(idTimeout)
              clearInterval(idInterval)
              reject(err)
            })
        }, 1000)
      }))
  }

  putConnector (connector) {
    return this.fetch('PUT', `konnectors/${connector.id}`, connector)
      .then(response => {
        return response.status === 200
          ? response.text()
          : Promise.reject(response)
      })
      .then((body) => {
        let connector = JSON.parse(body)
        this.updateConnector(connector)
        return connector
      })
  }

  importConnector (connector) {
    return this.fetch('POST', `konnectors/${connector.id}/import`, connector)
      .then(response => {
        return response.status === 200
          ? response
          : Promise.reject(response)
      })
  }

  startConnectorPoll (connectorId, timeout = 30000, interval = 500) {
    let endTime = Number(new Date()) + timeout

    let checkCondition = function (resolve, reject) {
      return this.fetch('GET', `konnectors/${connectorId}`)
        .then(response => response.text()).then(body => {
          let connector = JSON.parse(body)
          if (!connector.isImporting) {
            this.updateConnector(connector)
            resolve(connector)
          } else if (Number(new Date()) < endTime) {
            setTimeout(checkCondition, interval, resolve, reject)
          } else {
            this.updateConnector(connector)
            reject(new Error('polling timed out'))
          }
        })
    }.bind(this)
    return new Promise((resolve, reject) => {
      setTimeout(checkCondition, 500, resolve, reject)
    })
  }

  refreshFolders () {
    return this.fetch('GET', 'folders')
      .then(response => response.text()).then(body => {
        this.folders = JSON.parse(body)
        Promise.resolve()
      })
  }

  fetch (method, url, body) {
    const STACK_DOMAIN = '//' + document.querySelector('[role=application]').dataset.cozyDomain
    const STACK_TOKEN = document.querySelector('[role=application]').dataset.cozyToken
    let params = {
      method: method,
      credentials: 'include',
      headers: {
        'Accept': 'application/json',
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${STACK_TOKEN}`
      }
    }
    if (body) {
      params.body = JSON.stringify(body)
    }
    return fetch(`${STACK_DOMAIN}${url}`, params)
      .then(response => {
        let data
        const contentType = response.headers.get('content-type')
        if (contentType && contentType.indexOf('json') >= 0) {
          data = response.json()
        } else {
          data = response.text()
        }

        return (response.status === 200 || response.status === 202 || response.status === 204)
          ? data
          : data.then(Promise.reject.bind(Promise))
      })
  }
}

export class Provider extends Component {
  getChildContext () {
    return { store: this.store }
  }

  constructor (props, context) {
    super(props, context)
    this.store = props.store
  }

  render ({children}) {
    return (children && children[0]) || null
  }
}
