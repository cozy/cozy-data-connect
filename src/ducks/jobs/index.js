import { fetchTriggerJobs } from 'redux-cozy-client'

// CRUD

export const fetchKonnectorJobs = () => fetchTriggerJobs('jobs', 'konnector')
