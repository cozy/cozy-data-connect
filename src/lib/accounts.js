/* accounts lib ready to be added to cozy-client-js */
import * as realtime from './realtime'

export const ACCOUNTS_DOCTYPE = 'io.cozy.accounts'

export function subscribeAll(cozy) {
  return realtime.subscribeAll(cozy, ACCOUNTS_DOCTYPE)
}
