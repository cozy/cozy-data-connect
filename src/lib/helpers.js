/* global cozy */

export const getAccountName = account => {
  if (!account) return null
  if (account.auth) {
    return account.auth.accountName || getAccountLogin(account) || account._id
  } else {
    return account._id
  }
}

export const getAccountLogin = account => {
  if (account && account.auth) {
    return account.auth.login || account.auth.identifier || account.auth.email
  }
}

export const redirectToStore = () => {
  cozy.client.intents.redirect('io.cozy.konnectors', { type: 'konnector' })
}
