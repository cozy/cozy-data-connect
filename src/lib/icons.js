// Fallback to get the item icon and avoid error if not found
// with a possible default icon

export const getKonnectorIcon = konnector => {
  if (konnector.icon) {
    return require(`../assets/konnectors/${konnector.icon}`)
  }
  const slug = konnector.slug
  let icon = ''
  const extensions = ['.svg', '.png', '.gif', '.jpg']
  for (const ext of extensions) {
    try {
      icon = require(`../assets/konnectors/${slug}${ext}`)
      break
    } catch (e) {} // eslint-disable-line
  }
  if (!icon) {
    icon = require('../assets/konnectors/default.svg')
  }
  return icon
}
