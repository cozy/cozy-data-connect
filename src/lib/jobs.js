/* jobs lib ready to be added to cozy-client-js */
import * as realtime from './realtime'

const JOBS_DOCTYPE = 'io.cozy.jobs'

function decode(job) {
  // Retrieve message properties directly into job
  return { ...job, ...job.message }
}

export function subscribe(cozy, job) {
  return realtime.subscribe(cozy, JOBS_DOCTYPE, job, decode)
}

export function subscribeAll(cozy) {
  return realtime.subscribeAll(cozy, JOBS_DOCTYPE, decode)
}
