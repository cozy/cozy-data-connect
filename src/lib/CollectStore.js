/* global fetch */
/* global cozy */
import { Component } from 'react'

import * as accounts from './accounts'
import * as konnectors from './konnectors'

const INSTALL_TIMEOUT = 120 * 1000

export default class CollectStore {
  constructor (connectors, folders, context) {
    this.listener = null
    this.connectors = connectors
    this.folders = folders
    this.useCases = require(`../contexts/${context}/index`).useCases
  }

  subscribeTo (connectorId, listener) {
    this.listener = listener
    return this.find(c => c.id === connectorId)
  }

  unsubscribe () {
    this.listener = null
  }

  updateConnector (connector) {
    if (connector) {
      this.connectors = this.connectors.map(
        c => c.slug === connector.slug ? Object.assign({}, c, connector) : c
      )
    }
    return this.connectors.find(k => k.slug === connector.slug)
  }

  getCategories () {
    return this.connectors.map(a => a.category).filter((cat, idx, all) => all.indexOf(cat) === idx)
  }

  getUseCases () {
    return this.useCases
  }

  find (cb) {
    return this.connectors.find(cb)
  }

  findConnected () {
    return this.connectors.filter(c => c.accounts.length !== 0)
  }

  findByCategory ({filter}) {
    return filter === 'all' ? this.connectors
      : this.connectors.filter(c => c.category === filter)
  }

  findByUseCase (slug) {
    let useCase = this.useCases.find(u => u.slug === slug)
    return useCase.connectors.map(c1 => this.find(c2 => c1.slug === c2.slug))
  }

  // Fetch all accounts and updates their matching connectors
  fetchAllAccounts () {
    return accounts.getAllAccounts(cozy.client, this.accountsIndex)
      .then(accounts => {
        let accObject = {}
        accounts.forEach(a => {
          if (!accObject[a.account_type]) accObject[a.account_type] = []
          accObject[a.account_type].push(a)
        })
        this.connectors.forEach(c => {
          c.accounts = accObject[c.slug] || []
        })
      })
  }

  // Account connection workflow, see
  // https://github.com/cozy/cozy-stack/blob/master/docs/konnectors_workflow_example.md
  connectAccount (konnector, account, folderPath) {
    // return object to store all business object implied in the connection
    const connection = {}
    // detect oauth case
    const isOAuth = !!account.oauth

    // 1. Create folder, will be replaced by an intent or something else
    return cozy.client.files.createDirectoryByPath(folderPath)
      // 2. Create account
      .then(folder => {
        connection.folder = folder
        if (isOAuth) {
          const newAttributes = {
            folderId: folder._id,
            status: 'PENDING'
          }
          return accounts.update(cozy.client, account, Object.assign({}, account, newAttributes))
        } else {
          return accounts.create(cozy.client, konnector, account.auth, folder)
        }
      })
      // 3. Konnector installation
      .then(account => {
        connection.account = account
        return konnectors.install(cozy.client, konnector, INSTALL_TIMEOUT)
      })
      // 4. Add account to konnector
      .then(konnector => {
        return konnectors.addAccount(cozy.client, konnector, connection.account)
      })
      // 5. Add permissions to folder for konnector
      .then(konnector => {
        connection.konnector = konnector
        this.updateConnector(konnector)
        return konnectors.addFolderPermission(cozy.client, konnector, connection.folder._id)
      })
      // 6. Reference konnector in folder
      .then(permission => {
        connection.permission = permission
        return cozy.client.data.addReferencedFiles(connection.konnector, connection.folder._id)
      })
      // 7. Run a job for the konnector
      .then(() => konnectors.run(
        cozy.client,
        connection.konnector,
        connection.account))
      // 8. Creates trigger
      .then(job => {
        connection.job = job
        return cozy.client.fetchJSON('POST', '/jobs/triggers', {
          data: {
            attributes: {
              type: '@cron',
              arguments: '0 0 0 * * *',
              worker: 'konnector',
              worker_arguments: {
                konnector: connection.konnector._id,
                account: connection.account._id,
                folderToSave: connection.folder._id
              }
            }
          }
        })
      })
      .then(() => connection)
  }

  fetchAccounts (accountType, index) {
    if (!index && this.accountsIndex) index = this.accountsIndex
    return accounts.getAccountsByType(cozy.client, accountType, index)
  }

  /**
   * updateAccount : updates an account in a connector in DB with new values
   * @param {Object} connector The connector to update
   * @param {Object} account   The account to update
   * @param {Object} values    The new values of the updated account
   * @returns {Object} The up to date connector
   */
  updateAccount (connector, account, values) {
    // Save the previous state
    const previousAccount = Object.assign({}, account)

    // Update account data
    account.auth.login = values.login
    account.auth.password = values.password

    return accounts.update(cozy.client, previousAccount, account)
    .then(updatedAccount => {
      const accountIndex = this.findAccountIndexById(connector.accounts, account._id)
      // Updates the _rev value of the account in the connector
      connector.accounts[accountIndex] = updatedAccount
      this.updateConnector(connector)
    })
    .catch((error) => {
      return Promise.reject(error)
    })
  }

  /**
   * findAccountIndexById : returns the account index in an array of accounts, based on its id.
   * @param {array}    accounts Array of accounts
   * @param {string}   id       Id of the account we look for
   * @return {integer} The position of the account with the looked for id in the array
   */
  findAccountIndexById (accounts, id) {
    let foundIndex = -1
    accounts.forEach((account, index) => {
      if (account._id === id) {
        foundIndex = index
      }
    })
    return foundIndex
  }

  synchronize (connectorId) {
    let connector = this.find(c => c.id === connectorId)
    return this.importConnector(connector)
      .then(() => this.startConnectorPoll(connector.id))
  }

  deleteAccount (konnector, account) {
    konnector = this.connectors.find(c => c._id === konnector._id)
    konnector.accounts.splice(konnector.accounts.indexOf(account), 1)

    return accounts._delete(cozy.client, account)
      .then(() => konnectors.unlinkFolder(cozy.client, konnector, account.folderId))
      .then(() => this.updateConnector(konnector))
  }

  manifestToKonnector (manifest) {
    return manifest
  }

  // get properties from installed konnector or remote manifest
  fetchKonnectorInfos (slug) {
    return this.getInstalledConnector(slug)
      .then(konnector => {
        if (!konnector) {
          konnector = this.connectors.find(k => k.slug === slug)
        }

        return konnector ? konnectors.fetchManifest(cozy.client, konnector.repo)
          .then(this.manifestToKonnector)
          .catch(error => {
            console.warn && console.warn(`Cannot fetch konnector's manifest (Error ${error.status})`)
            return konnector
          }) : null
      })
      .then(konnector => konnector ? this.updateConnector(konnector) : null)
  }

  getInstalledConnector (slug) {
    return konnectors.findBySlug(cozy.client, slug)
  }

  putConnector (connector) {
    return this.fetch('PUT', `konnectors/${connector.id}`, connector)
      .then(response => {
        return response.status === 200
          ? response.text()
          : Promise.reject(response)
      })
      .then((body) => {
        let connector = JSON.parse(body)
        this.updateConnector(connector)
        return connector
      })
  }

  importConnector (connector) {
    return this.fetch('POST', `konnectors/${connector.id}/import`, connector)
      .then(response => {
        return response.status === 200
          ? response
          : Promise.reject(response)
      })
  }

  startConnectorPoll (connectorId, timeout = 30000, interval = 500) {
    let endTime = Number(new Date()) + timeout

    let checkCondition = function (resolve, reject) {
      return this.fetch('GET', `konnectors/${connectorId}`)
        .then(response => response.text()).then(body => {
          let connector = JSON.parse(body)
          if (!connector.isImporting) {
            this.updateConnector(connector)
            resolve(connector)
          } else if (Number(new Date()) < endTime) {
            setTimeout(checkCondition, interval, resolve, reject)
          } else {
            this.updateConnector(connector)
            reject(new Error('polling timed out'))
          }
        })
    }.bind(this)
    return new Promise((resolve, reject) => {
      setTimeout(checkCondition, 500, resolve, reject)
    })
  }

  refreshFolders () {
    return this.fetch('GET', 'folders')
      .then(response => response.text()).then(body => {
        this.folders = JSON.parse(body)
        Promise.resolve()
      })
  }

  fetch (method, url, body) {
    const STACK_DOMAIN = '//' + document.querySelector('[role=application]').dataset.cozyDomain
    const STACK_TOKEN = document.querySelector('[role=application]').dataset.cozyToken
    let params = {
      method: method,
      credentials: 'include',
      headers: {
        'Accept': 'application/json',
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${STACK_TOKEN}`
      }
    }
    if (body) {
      params.body = JSON.stringify(body)
    }
    return fetch(`${STACK_DOMAIN}${url}`, params)
      .then(response => {
        let data
        const contentType = response.headers.get('content-type')
        if (contentType && contentType.indexOf('json') >= 0) {
          data = response.json()
        } else {
          data = response.text()
        }

        return (response.status === 200 || response.status === 202 || response.status === 204)
          ? data
          : data.then(Promise.reject.bind(Promise))
      })
  }

  createIntentService (intent, window) {
    return cozy.client.intents.createService(intent, window)
  }
}

export class Provider extends Component {
  getChildContext () {
    return { store: this.store }
  }

  constructor (props, context) {
    super(props, context)
    this.store = props.store
  }

  render ({children}) {
    return (children && children[0]) || null
  }
}
