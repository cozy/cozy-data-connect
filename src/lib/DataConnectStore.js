/* global fetch */
/* global cozy */
import { Component } from 'react'

import * as accounts from './accounts'
import * as konnectors from './konnectors'

const INSTALL_TIMEOUT = 120
const KONNECTOR_STATE_READY = 'ready'

export default class DataConnectStore {
  constructor (connectors, folders, context) {
    this.listener = null
    this.connectors = connectors
    this.folders = folders
    this.useCases = require(`../contexts/${context}/index`).useCases
  }

  subscribeTo (connectorId, listener) {
    this.listener = listener
    return this.find(c => c.id === connectorId)
  }

  unsubscribe () {
    this.listener = null
  }

  updateConnector (connector) {
    if (connector) {
      this.connectors = this.connectors.map(
        c => c.slug === connector.slug ? Object.assign({}, c, connector) : c
      )
      if (this.listener) {
        this.listener(this.find(c => c.slug === connector.slug))
      }
    }
    return this.connectors.find(k => k.slug === connector.slug)
  }

  getCategories () {
    return this.connectors.map(a => a.category).filter((cat, idx, all) => all.indexOf(cat) === idx)
  }

  getUseCases () {
    return this.useCases
  }

  find (cb) {
    return this.connectors.find(cb)
  }

  findConnected () {
    return this.connectors.filter(c => c.accounts.length !== 0)
  }

  findByCategory ({filter}) {
    return filter === 'all' ? this.connectors
      : this.connectors.filter(c => c.category === filter)
  }

  findByUseCase (slug) {
    let useCase = this.useCases.find(u => u.slug === slug)
    return useCase.connectors.map(c1 => this.find(c2 => c1.slug === c2.slug))
  }

  getKonnectorFolder (konnector, base = '/') {
    base = base.charAt(base.length) === '/'
      ? base.substr(0, base.length - 1)
        : base

    base = base.charAt(0) === '/'
      ? base
        : `/${base}`

    const folderPath = `${base}/${konnector.name}`
    return cozy.client.files.createDirectoryByPath(folderPath)
  }

  connectAccount (konnector, account, folder) {
    const result = account.id
      // TODO: replace by updateAccount
      ? Promise.resolve(account)
        : this.addAccount(konnector, account.values)
    result
    .then(account => {
      return cozy.client.fetchJSON('PATCH', konnector.links.permissions, {
        data: {
          id: konnector.links.permissions,
          type: 'io.cozy.permissions',
          attributes: {
            type: 'app',
            source_id: konnector._id,
            permissions: {
              saveFolder: {
                type: 'io.cozy.files',
                values: [folder._id]
              }
            }
          }
        }
      })
      .then(account => {
        // add a reference to the folder in the konnector
        return cozy.client.fetchJSON('POST', `/files/${folder._id}/relationships/referenced_by`, {
          data: {
            type: 'io.cozy.konnectors',
            id: konnector._id
          }
        })
      })
      .then(() => konnectors.run(cozy.client, konnector.slug, account._id, folder._id))
      .then(() => {
        // create a trigger to run the job every weeks (default value)
        return cozy.client.fetchJSON('POST', '/jobs/triggers', {
          data: {
            attributes: {
              type: '@cron',
              arguments: '0 0 0 * * *',
              worker: 'konnector',
              worker_arguments: {
                konnector: konnector._id,
                account: account._id,
                folderToSave: folder._id
              }
            }
          }
        })
      })
    })

    return result
  }

  isInstalled (konnector) {
    return konnector.state && konnector.state === KONNECTOR_STATE_READY
  }

  addAccount (konnector, auth) {
    return accounts.create(cozy.client, konnector, auth)
      .then(account => {
        const installationPromise = this.isInstalled(konnector)
          ? Promise.resolve(konnector)
            : konnectors.install(cozy.client, konnector.slug, konnector.repo, INSTALL_TIMEOUT * 1000)

        return installationPromise
          .then(konnectorResult => {
            konnector.links = konnectorResult.links
            return konnectors.addAccount(cozy.client, konnectorResult, account)
          })
      })
  }

  fetchAccounts (accountType, index) {
    if (!index && this.accountsIndex) index = this.accountsIndex
    return accounts.getAccountsByType(cozy.client, accountType, index)
  }

  updateAccount (connectorId, accountIdx, values) {
    let connector = this.find(c => c.id === connectorId)
    connector.accounts[accountIdx] = values
    return this.putConnector(connector)
  }

  synchronize (connectorId) {
    let connector = this.find(c => c.id === connectorId)
    return this.importConnector(connector)
      .then(() => this.startConnectorPoll(connector.id))
  }

  deleteAccount (connectorId, accountIdx) {
    let connector = this.find(c => c.id === connectorId)
    connector.accounts.splice(accountIdx, 1)
    return this.importConnector(connector)
      .then(() => this.updateConnector(connector))
  }

  manifestToKonnector (manifest) {
    return manifest
  }

  // get properties from installed konnector or remote manifest
  fetchKonnectorInfos (slug) {
    return this.getInstalledConnector(slug)
      .then(konnector => {
        if (!konnector) {
          konnector = this.connectors.find(k => k.slug === slug)
        }

        return konnector ? konnectors.fetchManifest(cozy.client, konnector.repo)
          .then(this.manifestToKonnector)
          .catch(error => {
            console.warn && console.warn(`Cannot fetch konnector's manifest (Error ${error.status})`)
            return konnector
          }) : null
      })
      .then(konnector => konnector ? this.updateConnector(konnector) : null)
  }

  getInstalledConnector (slug) {
    return konnectors.findBySlug(cozy.client, slug)
  }

  putConnector (connector) {
    return this.fetch('PUT', `konnectors/${connector.id}`, connector)
      .then(response => {
        return response.status === 200
          ? response.text()
          : Promise.reject(response)
      })
      .then((body) => {
        let connector = JSON.parse(body)
        this.updateConnector(connector)
        return connector
      })
  }

  importConnector (connector) {
    return this.fetch('POST', `konnectors/${connector.id}/import`, connector)
      .then(response => {
        return response.status === 200
          ? response
          : Promise.reject(response)
      })
  }

  startConnectorPoll (connectorId, timeout = 30000, interval = 500) {
    let endTime = Number(new Date()) + timeout

    let checkCondition = function (resolve, reject) {
      return this.fetch('GET', `konnectors/${connectorId}`)
        .then(response => response.text()).then(body => {
          let connector = JSON.parse(body)
          if (!connector.isImporting) {
            this.updateConnector(connector)
            resolve(connector)
          } else if (Number(new Date()) < endTime) {
            setTimeout(checkCondition, interval, resolve, reject)
          } else {
            this.updateConnector(connector)
            reject(new Error('polling timed out'))
          }
        })
    }.bind(this)
    return new Promise((resolve, reject) => {
      setTimeout(checkCondition, 500, resolve, reject)
    })
  }

  refreshFolders () {
    return this.fetch('GET', 'folders')
      .then(response => response.text()).then(body => {
        this.folders = JSON.parse(body)
        Promise.resolve()
      })
  }

  fetch (method, url, body) {
    const STACK_DOMAIN = '//' + document.querySelector('[role=application]').dataset.cozyDomain
    const STACK_TOKEN = document.querySelector('[role=application]').dataset.cozyToken
    let params = {
      method: method,
      credentials: 'include',
      headers: {
        'Accept': 'application/json',
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${STACK_TOKEN}`
      }
    }
    if (body) {
      params.body = JSON.stringify(body)
    }
    return fetch(`${STACK_DOMAIN}${url}`, params)
      .then(response => {
        let data
        const contentType = response.headers.get('content-type')
        if (contentType && contentType.indexOf('json') >= 0) {
          data = response.json()
        } else {
          data = response.text()
        }

        return (response.status === 200 || response.status === 202 || response.status === 204)
          ? data
          : data.then(Promise.reject.bind(Promise))
      })
  }

  createIntentService (intent, window) {
    return cozy.client.intents.createService(intent, window)
  }
}

export class Provider extends Component {
  getChildContext () {
    return { store: this.store }
  }

  constructor (props, context) {
    super(props, context)
    this.store = props.store
  }

  render ({children}) {
    return (children && children[0]) || null
  }
}
