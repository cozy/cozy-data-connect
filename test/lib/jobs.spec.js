/* eslint-env jest */
import * as jobs from '../../src/lib/jobs'

const jobsMock = [
  {
    attributes: {
      _id: 'b7671bbf53e9687e27487b617a0202fd',
      _rev: '3-0138e101246dfa67f80dfaa801ba4dc9',
      domain: 'cozy.tools:8080',
      worker: 'konnector',
      message: {
        konnector: 'trainline',
        account: 'b7671bbf53e9687e27487b617a01ffff',
        folder_to_save: 'b7671bbf53e9687e27487b617a003b6e'
      },
      options: {
        max_exec_count: 1,
        max_exec_time: 0,
        timeout: 0
      },
      state: 'running',
      queued_at: '2017-10-04T15:59:59.737529564+02:00',
      started_at: '2017-10-04T16:00:01.767102968+02:00'
    }
  },
  {
    attributes: {
      _id: 'b7671bbf53e9687e27487b617a0224d7',
      _rev: '3-fed0f44c3a6ccabfe2c3c7b9bef9a0be',
      domain: 'cozy.tools:8080',
      worker: 'konnector',
      message: {
        konnector: 'trainline',
        account: 'b7671bbf53e9687e27487b617a02248c',
        folder_to_save: 'b7671bbf53e9687e27487b617a003b6e'
      },
      options: {
        max_exec_count: 1,
        max_exec_time: 0,
        timeout: 0
      },
      state: 'done',
      queued_at: '2017-10-04T17:06:09.415785346+02:00',
      started_at: '2017-10-04T17:06:09.444339433+02:00'
    }
  }
]

// just to tests calling, results are tested in cozy-client-js
let cozyMock = {
  fetchJSON: jest.fn(() => Promise.resolve(jobsMock))
}

beforeEach(() => {
  // clear mocks calls
  cozyMock.fetchJSON.mockClear()
})

describe('jobs library', () => {
  it('does not ignore old jobs', () => {
    return jobs.findQueuedOrRunning(cozyMock).then(jobs => {
      expect(jobs).toMatchSnapshot()
    })
  })

  it('ignores oldest jobs', () => {
    return jobs
      .findQueuedOrRunning(cozyMock, new Date(Date.UTC(2017, 9, 4, 14)))
      .then(jobs => {
        expect(jobs).toMatchSnapshot()
      })
  })
})
