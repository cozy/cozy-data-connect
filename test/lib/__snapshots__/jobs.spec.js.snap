// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jobs library does not ignore old jobs 1`] = `
Array [
  Object {
    "_id": "b7671bbf53e9687e27487b617a0202fd",
    "_rev": "3-0138e101246dfa67f80dfaa801ba4dc9",
    "account": "b7671bbf53e9687e27487b617a01ffff",
    "domain": "cozy.tools:8080",
    "folder_to_save": "b7671bbf53e9687e27487b617a003b6e",
    "konnector": "trainline",
    "message": Object {
      "account": "b7671bbf53e9687e27487b617a01ffff",
      "folder_to_save": "b7671bbf53e9687e27487b617a003b6e",
      "konnector": "trainline",
    },
    "options": Object {
      "max_exec_count": 1,
      "max_exec_time": 0,
      "timeout": 0,
    },
    "queued_at": "2017-10-04T15:59:59.737529564+02:00",
    "started_at": "2017-10-04T16:00:01.767102968+02:00",
    "state": "running",
    "worker": "konnector",
  },
  Object {
    "_id": "b7671bbf53e9687e27487b617a0224d7",
    "_rev": "3-fed0f44c3a6ccabfe2c3c7b9bef9a0be",
    "account": "b7671bbf53e9687e27487b617a02248c",
    "domain": "cozy.tools:8080",
    "folder_to_save": "b7671bbf53e9687e27487b617a003b6e",
    "konnector": "trainline",
    "message": Object {
      "account": "b7671bbf53e9687e27487b617a02248c",
      "folder_to_save": "b7671bbf53e9687e27487b617a003b6e",
      "konnector": "trainline",
    },
    "options": Object {
      "max_exec_count": 1,
      "max_exec_time": 0,
      "timeout": 0,
    },
    "queued_at": "2017-10-04T17:06:09.415785346+02:00",
    "started_at": "2017-10-04T17:06:09.444339433+02:00",
    "state": "done",
    "worker": "konnector",
  },
]
`;

exports[`jobs library ignores oldest jobs 1`] = `
Array [
  Object {
    "_id": "b7671bbf53e9687e27487b617a0224d7",
    "_rev": "3-fed0f44c3a6ccabfe2c3c7b9bef9a0be",
    "account": "b7671bbf53e9687e27487b617a02248c",
    "domain": "cozy.tools:8080",
    "folder_to_save": "b7671bbf53e9687e27487b617a003b6e",
    "konnector": "trainline",
    "message": Object {
      "account": "b7671bbf53e9687e27487b617a02248c",
      "folder_to_save": "b7671bbf53e9687e27487b617a003b6e",
      "konnector": "trainline",
    },
    "options": Object {
      "max_exec_count": 1,
      "max_exec_time": 0,
      "timeout": 0,
    },
    "queued_at": "2017-10-04T17:06:09.415785346+02:00",
    "started_at": "2017-10-04T17:06:09.444339433+02:00",
    "state": "done",
    "worker": "konnector",
  },
]
`;
